### PR TITLE
Implement Copy button

### LINKED ISSUE #452: Copy button
It's a pain to select and copy the text every time

### PR DESCRIPTION
**Pull request by kota-yata**
close #452
image

Sorry for sudden PR. I would be glad if you could merge.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 .idea
+node_modules
```
Comment (mozfreddyb) [inline .gitignore]: please add newline at the end of the file

File: app.js [hunk 2]
```diff
@@ -34,6 +37,18 @@ async function hashText(message) {
...
+function copyText(text) {
+  const str = text;
+
+  function listener(e) {
+    e.clipboardData.setData("text/plain", str);
+    e.preventDefault();
+    document.removeEventListener("copy", listener);
+  }
+  document.addEventListener("copy", listener);
+  document.execCommand("copy");
```
Comment (mozfreddyb) [inline app.js]: I haven't worked with clipboard APIs before. Which browsers did you test? :-)

File: style.css [hunk 3]
```diff
@@ -154,6 +156,9 @@ footer {
   background-color: #a9a9a9;
   color: #fff;
 }
+#sriCopy {
+  width: 50px;
```
Comment (mozfreddyb) [inline style.css]: I think we're trying to avoid the pixel unit for sizing. Can you use rem, to be consistent with other platforms??

**Review by mozfreddyb — CHANGES_REQUESTED** [review]
looks good! almost ready.

File: app.js (hunk 2 above)
Comment (kota-yata) [reply app.js]: I test this on Chrome. I'm not sure it works on IE, but I think it works correctly on the other browser.

Comment (kota-yata) [reply app.js]: I tested on Edge and Safari now, and doesn't work on Safari. I will fix this now.

Comment (mozfreddyb) [reply app.js]: Firefox support is also kinda important to us here at Mozilla :slightly_smiling_face:

**Comment by kota-yata** [comment]
Clipboard API doesn't work on Safari because Safari doesn't support Permission API, and other ways I tried also doesn't work on Safari. Finally I found [clipboard.js](https://clipboardjs.com/) and this works on any browser.

Of course, it works on Firefox completely 🦊

**Comment by mozfreddyb** [comment]
I'm hopeful this can be done without a third party library. the code at https://github.com/github/clipboard-copy-element/blob/main/src/clipboard.ts seems very concise (though it's typescript and needs a bit of a re-implementation :))

**Comment by kota-yata** [comment]
This is really good code 😮 
I made some changes and it works on any browser!!

File: app.js [hunk 4]
```diff
@@ -59,6 +115,15 @@ async function formSubmit(event) {
         url
       )}&quot;</span> <span style="color: #ffa07a">integrity=</span><span style="color:#abe338">&quot;${integrityMetadata}&quot;</span> <span style="color: #ffa07a">crossorigin=</span><span style="color:#abe338">&quot;anonymous&quot;</span><span style="color: #ffa07a">&gt;&lt;/script&gt;</span>`;
 
+      resultDiv.innerText = scriptEl;
```
Comment (mozfreddyb) [inline app.js]: I think your merge went wrong. The latest commit has resultDiv.innerHTML.

File: app.js [hunk 5]
```diff
@@ -35,6 +38,59 @@ async function hashText(message) {
...
+  }
+
+  const node = createNode(text);
+
+  body.appendChild(node);
+  copyNode(node);
+  body.removeChild(node);
+  return Promise.resolve();
+}
+
```
Comment (mozfreddyb) [inline app.js]: Please move these functions into a second JS file, clipboard.js and add that to the HTML as a `script` element to the head.
While you're at it, please also move the other `<script>` element to the head and make both `defer`.

File: style.css [hunk 6]
```diff
@@ -131,7 +131,9 @@ footer {
   padding: 0 2%;
 }
 
-#sriApp input[type="submit"] {
+#sriApp input[type="submit"],
+
+#sriCopy {
```
Comment (mozfreddyb) [inline style.css:140]: Please put it on the same line like so:
`#sriApp input[type="submit"], #sriCopy {`

**Review by mozfreddyb — CHANGES_REQUESTED** [review]
Thanks, that looks much better.
Please address my comments above and, if you can, put this on your github pages, so I can do some testing. (You'll need to fork into _another_ branch that doesn't have the CNAME file)

File: app.js [hunk 7]
```diff
@@ -59,6 +62,16 @@ async function formSubmit(event) {
...
       )}&quot;</span> <span style="color: #ffa07a">integrity=</span><span style="color:#abe338">&quot;${integrityMetadata}&quot;</span> <span style="color: #ffa07a">crossorigin=</span><span style="color:#abe338">&quot;anonymous&quot;</span><span style="color: #ffa07a">&gt;&lt;/script&gt;</span>`;
 
+      resultDiv.innerHTML = scriptEl;
+      const copyButton = `<button id="sriCopy">Copy</button>`;
+
+      resultDiv.insertAdjacentHTML('afterend', copyButton);
+      const sriCopy = document.getElementById("sriCopy");
+
+      sriCopy.addEventListener("click", () => {
+        // eslint-disable-next-line no-undef
```
Comment (mozfreddyb) [inline app.js]: I don't think this is ndded anymore

File: app.js [hunk 8]
```diff
@@ -59,6 +62,16 @@ async function formSubmit(event) {
...
+        copyText(resultDiv.innerText);
+      });
       resultDiv.classList.add("is-active");
       resultDiv.innerHTML = scriptEl;
```
Comment (mozfreddyb) [inline app.js]: This and line 65 are duplicated

File: clipboard.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+  range.selectNodeContents(node);
+  selection.addRange(range);
+
+  document.execCommand('copy');
+  selection.removeAllRanges();
+  return Promise.resolve();
+}
+
+// eslint-disable-next-line no-unused-vars
```
Comment (mozfreddyb) [inline clipboard.js:32]: why is this needed?

File: index.html [hunk 10]
```diff
@@ -39,6 +39,8 @@
     <meta property="twitter:title" content="SRI Hash Generator">
     <meta property="twitter:description" content="Generate Subresource Integrity (SRI) attributes, easily and fast">
     <meta property="twitter:image" content="">
+    <script src="clipboard.js" defer></script>
+    <script src="app.js" defer></script>
```
Comment (mozfreddyb) [inline index.html:43]: thank you!

**Comment by mozfreddyb** [comment]
almost :-)

File: app.js (hunk 7 above)
Comment (kota-yata) [reply app.js]: This is needed because there is no ```copyText()``` in app.js. Is there any idea to avoid this error without making eslint disable?

**Comment by kota-yata** [comment]
I made GitHub pages for test, so please check this :)
[https://kota-yata.github.io/srihash-gh/](https://kota-yata.github.io/srihash-gh/)

File: app.js (hunk 7 above)
Comment (kota-yata) [reply app.js]:
```javascript
sriCopy.addEventListener("click", () => {
   // eslint-disable-next-line no-undef
   copyText(resultDiv.innerText);
});
```

This function have to be in ```formSubmit()```. Otherwise I will get an error like so:
image
because when I load this page there is no ```sriCopy```.

Comment (mozfreddyb) [reply app.js]: The preferred way is to tell eslint about this global. In the very first line of the main js file, add
`/* global copyText */` and other things that you need exposed.

**Review by mozfreddyb — CHANGES_REQUESTED** [review]
Please use the `/* global .. */`  comment as mentioned above and we should be almost ready to land.

File: clipboard.js (hunk 9 above)
Comment (kota-yata) [reply clipboard.js:37]: How can I avoid this eslint-disable? ```/* global ... */``` doesn't work for function definitions.

**Comment by kota-yata** [comment]
I think this is ready to land...

**Comment by mozfreddyb** [comment]
@kota-yata sorry for the generally slow response rate from my side. Would you be willing to resolve the conflicts and clean up the git history of this branch a bit? Looking forward to what you come up with :-)

**Comment by kota-yata** [comment]
Since I've never cleaned up git history, I actually don't know how to do it. Should I do ```git rebase -i 1a4eacb```?

**Comment by mozfreddyb** [comment]
Yes. `rebase -i` is what I would use. Maybe you will find a graphical tool that can help you combine related commits. Thanks!

File: app.js [hunk 11]
```diff
@@ -95,7 +100,7 @@ async function formSubmit(event) {
       const type = response.headers.get("content-type");
       const contentType = parseContentType(type);
       const text = await response.text();
-
+
```
Comment (mozfreddyb) [inline app.js]: Please remove extra whitespace

File: clipboard.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+function createNode(text) {
+  const node = document.createElement('pre');
+
+  node.style.width = '1px';
+  node.style.height = '1px';
+  node.style.position = 'fixed';
+  node.style.top = '5px';
+  node.textContent = text;
+  return node;
```
Comment (mozfreddyb) [inline clipboard.js:6]: Can you assign a `class` to the node and move the styling into the CSS file?

Comment (mozfreddyb) [reply clipboard.js:6]: Can we use an element that has "display: none" instead?

**Review by mozfreddyb — CHANGES_REQUESTED** [review]
We're getting very close. Thanks @kota-yata!

File: app.js (hunk 11 above)
Comment (kota-yata) [reply app.js]: Isn't this line break set by eslint ```newline-after-var``` ?

Comment (mozfreddyb) [reply app.js]: The newline is OK, but it seems to contain a bunch of spaces (`...await response.text();` linebreak, space, space, space ... linebreak). Instead, we want (`...await response.text();`, linebreak, linebreak).

We call these "extra whitespaces" or trailing whitespace

**Comment by mozfreddyb** [comment]
This seems to work great in my testing. Thank you so much! :+1:

**Comment by kota-yata** [comment]
Although rebasing was not in time, Thanks!

**Comment by mozfreddyb** [comment]
don't worry, I squashed it all together.
